File: 283.MoveZeroes/283.cpp
/*
 * swap non-zero number to tail of processed, O(n)
 */
#include <algorithm> // swap
#include "catch.hpp"
using std::vector;
using std::swap;

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        for(int scan = 0, tail = 0; scan < nums.size(); ++scan)
            if(nums[scan]) swap(nums[tail++], nums[scan]);
    }
};

TEST_CASE("moving zeroes", "[moveZeroes]") {
    Solution s;
    vector<int> a{0,1,0,3,12};
    vector<int> b{1,3,12,0,0};
    s.moveZeroes(a);
    CHECK(a == b);
}

File: 283.Move Zeroes/283.java
/*
 * swap non-zero number to tail of processed, O(n)
 */

class Solution {
    public void moveZeroes(int[] nums) {
        int tail = 0;
        for(int scan = 0; scan < nums.length; scan++)
            if(nums[scan] != 0) nums[tail++] = nums[scan];
        while(tail < nums.length) nums[tail++] = 0;
    }
    public static void main(String[] args) {
        Solution s = new Solution();
        int[] a = {0,1,0,3,12};
        s.moveZeroes(a);
        for(int i : a) System.out.println(i);
    }
}
